Source/FDWaveformView.swift

```diff
@@ -568,15 +576,23 @@ extension FDWaveformView: UIGestureRecognizerDelegate {
         else if doesAllowScrubbing {
             let rangeSamples = CGFloat(zoomSamples.count)
             let scrubLocation = min(max(recognizer.location(in: self).x, 0), frame.width)    // clamp location within the frame
-            highlightedSamples = 0 ..< Int((CGFloat(zoomSamples.startIndex) + rangeSamples * scrubLocation / bounds.width))
-            delegate?.waveformDidEndScrubbing?(self)
+            scrub(to: Int((CGFloat(zoomSamples.startIndex) + rangeSamples * scrubLocation / bounds.width)))
         }
     }
 
     @objc func handleTapGesture(_ recognizer: UITapGestureRecognizer) {
         if doesAllowScrubbing {
             let rangeSamples = CGFloat(zoomSamples.count)
-            highlightedSamples = 0 ..< Int((CGFloat(zoomSamples.startIndex) + rangeSamples * recognizer.location(in: self).x / bounds.width))
+            scrub(to: Int((CGFloat(zoomSamples.startIndex) + rangeSamples * recognizer.location(in: self).x / bounds.width)))
+        }
+    }
+    
+    func scrub(to sample: Int) {
+        if boundToScrub == .upper, let lowerBound = highlightedSamples?.lowerBound, sample > lowerBound {
+            highlightedSamples = lowerBound ..< sample
+            delegate?.waveformDidEndScrubbing?(self)
+        } else if let upperBound = highlightedSamples?.upperBound, sample < upperBound {
+            highlightedSamples = sample ..< upperBound
             delegate?.waveformDidEndScrubbing?(self)
         }
```

Comment:
Decided to move `delegate?.waveformDidEndScrubbing?(self)` out of `handleXYZGesture` and into here so that it only gets called when scrubbing to a valid location. If this isn't desired behaviour I can move it back again.